docs/upgrading/production-basics.md

```diff
@@ -0,0 +1,233 @@
+:::
+
+## Using Enums
+
+In the example above, all contract state is stored in one simple struct. Many real-world contracts are more complex, often having one struct referenced by another. For example, a [DAO](https://whiteboardcrypto.com/what-is-a-dao/) contract might look something like this:
+
+```rust
+#[derive(BorshSerialize, BorshDeserialize)]
+pub enum ProposalStatus {
+    Proposed,
+    Approved,
+    Rejected,
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct Proposal {
+    pub description: String,
+    pub status: ProposalStatus,
+}
+
+#[near_bindgen]
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct DAO {
+    pub proposals: LookupMap<u64, Proposal>,
+}
+```
+
+:::note
+For a more complete DAO example, check out [SputnikDAO](https://github.com/near-daos/sputnik-dao-contract/blob/317ea4fb1e6eac8064ef29a78054b0586a3406c3/sputnikdao2/src/lib.rs), [Flux](https://github.com/fluxprotocol/amm/blob/3def886a7fbd2df4ba28e18f67e6ab12cd2eee0b/dao/src/lib.rs), and [others](https://github.com/search?q=near+dao).
+:::
+
+Say you want to update the structure of `Proposal` but keep `DAO` unchanged.
+
+The first thing to note is that the contract could be storing a huge number of proposals, which makes it impossible to migrate all of them in one transaction due to [the gas limit](https://docs.near.org/docs/concepts/gas#thinking-in-gas). In an off-chain script, you could query the full state of the contract and update every single one of them via multiple transactions. But that may be prohibitively expensive, so you might opt to upgrade proposals to the new structure during the next interaction with them, rather than all at once (this disperses the upgrade cost to users of the contract).
+
+In either case, your contract can end up with proposals using the original structure and the new structure at the same time, and the `DAO` struct needs to know how to load both of them. How do you do that?
+
+Use [enums](https://doc.rust-lang.org/book/ch06-00-enums.html):
+
+```rust
+#[derive(BorshSerialize, BorshDeserialize)]
+pub enum ProposalStatus {
+    Proposed,
+    Approved,
+    Rejected,
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct ProposalV1 {
+    pub description: String,
+    pub status: ProposalStatus,
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct Proposal {
+    pub title: String,
+    pub description: String,
+    pub status: ProposalStatus,
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub enum UpgradableProposal {
+    V1(ProposalV1),
+    V2(Proposal),
+}
+
+impl From<UpgradableProposal> for Proposal {
+    fn from(proposal: UpgradableProposal) -> Self {
+        match proposal {
+            UpgradableAccount::V2(proposal) => proposal,
+            UpgradableAccount::V1(v1) => Proposal {
+                title: &v1.description[0..10],
```

Comment:
does this code exist somewhere? This line I don't think should compile?

Reply:
Thanks for checking; I did not try to compile this and I wasn't sure if my Rust syntax was correct. I saw this syntax [here](https://riptutorial.com/rust/example/4146/string-slicing); is there a better way you recommend to get the first N characters of a `String`?

Reply:
yeah it's not ideal to use slice indexing for strings because some characters can be more than 1 byte length (as is mentioned in the link) but the issue here is just that this is trying to use a `&str` when a `String` is needed. Adding `.to_string()` at the end will probably get it to compile, but it's unsafe as the string could not be utf8.

to safely index the first 10 characters you could do:
```rust
v1.description.get(..10).map(str::to_owned).unwrap_or_default(),
```

Which will just use an empty string if the description is not 10 characters long.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9d173464411960e65e6ef3cc707bd851